### PR TITLE
[Feat] A competitive Web Browsing agent

### PR DESCRIPTION
**Pull request by frankxu2004**
This PR aims at enabling a competitive browsing agent for #1470.

Now I transplanted the simplified demo agent used in WebArena in our agent hub.

To test, it works best with GPT-4 LLMs such as GPT-4o.

```
poetry run python ./opendevin/core/main.py -i 5 -t "tell me the usa's president using google search" -c BrowsingAgent -m gpt-4o-2024-05-13

```

### CODE REVIEW

**Comment by frankxu2004** [comment]
Example logs:

File: agenthub/browsing_agent/prompt.py [hunk 1]
```diff
@@ -0,0 +1,776 @@
...
+    ) -> None:
+        super().__init__(visible=visible)
+        self.max_line_diff = max_line_diff
+        self.header, self.diff_lines = diff(previous, new)
+        self.shrink_speed = shrink_speed
+        self.prefix = prefix
+
+    def shrink(self):
+        self.max_line_diff -= self.shrink_speed
+        self.max_line_diff = max(1, self.max_line_diff)
```
Comment (yufansong) [inline agenthub/browsing_agent/prompt.py:611]: I guess these parameters and formulas come from some papers? I will be appreciate if you can share the paper link.

**Review by yufansong — APPROVED** [review]
Leave some nits. Mostly LGTM. I would be appricate it if you can add more comments or simply elaborate your design and some parameter setting. Then other people can add more work on your codebase. I don't want to block our integration progress and AP it. I can help for some follow up refactor or nits if you have no time.

File: agenthub/browsing_agent/browsing_agent.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        try:
+            cur_cost = self.llm.completion_cost(response)
+        except Exception:
+            cur_cost = 0
+        self.cost_accumulator += cur_cost
+        logger.info(
+            'Cost: %.2f USD | Accumulated Cost: %.2f USD',
+            cur_cost,
+            self.cost_accumulator,
+        )
```
[the remark this answers is not in the document]
Comment (frankxu2004) [reply agenthub/browsing_agent/browsing_agent.py:149]: Sounds good!

**Comment by li-boxuan** [comment]
Sad, our project coverage reduced by 5.87%... let me thing if there's anything we could do to test this.

**Comment by li-boxuan** [comment]
I've made some progress in creating an integration test for this agent! Will create a PR in a day.

File: agenthub/browsing_agent/prompt.py [hunk 3]
```diff
@@ -0,0 +1,785 @@
...
+        self._prompt += '\n'.join(
+            [
+                f"""\
+ - [{msg['role']}] {msg['message']}"""
+                for msg in chat_messages
+            ]
+        )
+
+
+class SystemPrompt(PromptElement):
```
Comment (li-boxuan) [inline agenthub/browsing_agent/prompt.py:364]: @frankxu2004 this prompt (along with many other prompts in this file) seems unused? Is it by intention?

Comment (frankxu2004) [reply agenthub/browsing_agent/prompt.py:364]: Yes, basically this whole prompt.py file is not currently used. Currently the agent is a simplified version for ease of understanding. However I included here with the intention of incorporating a more complex agent using more comprehensive information as next steps. Here it's still useful as it provides others of building blocks of prompts and understanding what possible information to include as context for LLMs.

Comment (frankxu2004) [reply agenthub/browsing_agent/prompt.py:364]: These PRs are mostly for chasing the neurips paper deadline so not all features are implemented yet.

Comment (li-boxuan) [reply agenthub/browsing_agent/prompt.py:364]: I see, sounds fair. I am just having a bit trouble reproducing `poetry run python ./opendevin/core/main.py -i 5 -t "tell me the usa's president using google search" -c BrowsingAgent -m gpt-4o-2024-05-13`... I tried like 5 times and only succeeded once.

Comment (frankxu2004) [reply agenthub/browsing_agent/prompt.py:364]: that's a bit weird, what error are you seeing? do you have logs

Comment (frankxu2004) [reply agenthub/browsing_agent/prompt.py:364]: Currently the agent does not return AgentFinishAction, so to the eyes of the frame, it's always error in the end

Comment (li-boxuan) [reply agenthub/browsing_agent/prompt.py:364]: [logs.zip](https://github.com/OpenDevin/OpenDevin/files/15412109/logs.zip)

image

Basically, keep clicking without making progress

**Pull request merged**